core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -340,6 +385,45 @@ private[spark] class ExecutorAllocationManager(
     }
   }
 
+  /**
+   * Maximum number of executors to be removed per dra evaluation.
+   *
+   * This function limits the number of idle executors to be removed if the
+   * `streamingDRAFeatureEnabled` flag is enabled, otherwise it returns all the idle executors.
+   * The number of executors to be removed is based on
+   * the de-allocation ratio (`executorDeallocationRatio`) and timeout (`removeTime`).
+   * It helps in removing only few executors per evaluation cycle and helps in gradual removal of
+   * executors across micro-batches.
+   */
+  private def maxExecutorDeallocationsPerEvaluation(
+                                                     timedOutExecs : Seq[(String, Int)]
+                                                   ): Seq[(String, Int)] = {
+    if (!streamingDRAFeatureEnabled) {
+      timedOutExecs
+    }
+    else {
+      val currentTime = clock.nanoTime()
+
+      if (removeTime == NOT_SET) {
```

Comment:
Why do we remove all the timedOutExecs when removeTime is not set?  Doesn't this mean for the first evaluation of this function when removeTime is not set, we will remove all timeOutExecs and NOT adhere to the configured "executorDeallocationRatio"?

Reply:
Good catch. I will update it to `Seq.empty[(String, Int)]`. Every evaluation happens at a fixed rate of few hundred milliseconds. So all those executors will be evaluated in the next round